actions/actions_whitebox_test.go

```diff
@@ -0,0 +1,184 @@
+		require.NoError(t, err)
+		require.Len(t, changes, 1)
+		require.Equal(t, workitem.SystemState, changes[0].AttributeName)
+		require.Equal(t, "open", changes[0].NewValue)
+		require.Equal(t, "new", changes[0].OldValue)
+	})
+
+	s.T().Run("column changes", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "new", []interface{}{"bcid0"})
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.NoError(t, err)
+		require.Len(t, changes, 1)
+		fmt.Println(changes[0].OldValue)
+		fmt.Println(changes[0].NewValue)
+		require.Equal(t, workitem.SystemBoardcolumns, changes[0].AttributeName)
+		require.Equal(t, wiCopy.Fields[workitem.SystemBoardcolumns], changes[0].OldValue)
+		require.Equal(t, fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns], changes[0].NewValue)
+	})
+
+	s.T().Run("multiple changes", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0"})
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.NoError(t, err)
+		require.Len(t, changes, 2)
+		// we intentionally test the order here as the code under test needs
+		// to be expanded later, supporting more changes and this is an
+		// integrity test on the current impl.
+		require.Equal(t, workitem.SystemState, changes[0].AttributeName)
+		require.Equal(t, "new", changes[0].NewValue)
+		require.Equal(t, "open", changes[0].OldValue)
+		require.Equal(t, workitem.SystemBoardcolumns, changes[1].AttributeName)
+		require.Equal(t, fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns], changes[1].NewValue)
+		require.Equal(t, wiCopy.Fields[workitem.SystemBoardcolumns], changes[1].OldValue)
+	})
+
+	s.T().Run("new instance", func(t *testing.T) {
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{}
+		changes, err := fxt.WorkItems[0].ChangeSet(nil)
+		require.NoError(t, err)
+		require.Len(t, changes, 1)
+		require.Equal(t, workitem.SystemState, changes[0].AttributeName)
+		require.Equal(t, "new", changes[0].NewValue)
+		require.Nil(t, changes[0].OldValue)
+	})
+}
+
+func (s *ActionSuite) TestActionExecution() {
+	fxt := tf.NewTestFixture(s.T(), s.DB, tf.CreateWorkItemEnvironment(), tf.WorkItems(2))
+	userID := fxt.Identities[0].ID
+
+	s.T().Run("by Old New", func(t *testing.T) {
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		newVersion := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0", "bcid1"})
+		afterActionWI, changes, err := ExecuteActionsByOldNew(s.Ctx, s.GormDB, userID, fxt.WorkItems[0], newVersion, map[string]string{
```

Comment:
It does not make sense to work on beautifying test code that is already working and good enough. Yes, we could drag this on and on forever, but then we will never finish anything.

Reply:
If you introduce terminology like Action, Rule, ChangeSet, Context and so on you should reflect that in code somehow. How would I know that a ChangeSet is indexed an array of Changes and not a map of changes index by the field that was changed? 

This is not about beautifying code but doing it right the first time. And I don't care if a code is working or not. To this day we have issues in the UI exactly because of this mentality. Code was written because it fitted the story and nothing more. The moment something changes, it falls apart.

If things have their own type they then the Go documentation generation can do a much better job to generate a documentation that is readable and leads to more maintainable code. See the `workitem.Kind` for example:

```
$ go doc github.com/fabric8-services/fabric8-wit/workitem.Kind
type Kind string
    Kind is the kind of field type


const KindString Kind = "string" ...
func ConvertAnyToKind(any interface{}) (*Kind, error)
func ConvertStringToKind(k string) (*Kind, error)
func (k Kind) IsRelational() bool
func (k Kind) IsSimpleType() bool
func (k Kind) String() string
```

Now I agree that ` Kind is the kind of field type` is not a good description but the grouping is quite handy.

Btw. why Is the `convert.ChangeDetector` in the `convert` package? That has a completely different reason to exist.

That reminds me. Do you have an `actions/doc.go` added to describe your package?

Reply:
This is a skewed view of the world. Yes, code quality matters. But being reasonable and accepting an accountability to timelines has the same importance. We can go on and on and on to get the "perfect" code, but that does not bring anyone forward. The reason why we have issues is because we are overengineering stuff and because we are not reasonable on a lot of things.